File: Leetcode/0983 Minimum Cost For Tickets.ts
const durations = [1, 7, 30];

function mincostTickets(days: number[], costs: number[]): number {
    const cache: number[] = new Array(days.length).fill(null);

    function minCost(dayIndex: number) {
        if (dayIndex >= days.length) {
            return 0;
        }
        if (cache[dayIndex] != null) {
            return cache[dayIndex];
        }
        let cost = Infinity;
        for (let costIndex = 0; costIndex < costs.length; costIndex++) {
            let j = dayIndex;
            while (j < days.length && days[j] < days[dayIndex] + durations[costIndex]) {
                j++
            }
            cost = Math.min(cost, costs[costIndex] + minCost(j));
        }
        return cache[dayIndex] = cost;
    }

    return minCost(0);
}

File: Leetcode/0983 Minimum Cost For Tickets.py
durations = [1, 7, 30]

class Solution:
    def mincostTickets(self, days, costs):
        cache = [None for _ in days]

        def min_cost(day_index):
            if day_index >= len(days):
                return 0
                
            if cache[day_index] is not None:
                return cache[day_index]

            cache[day_index] = float('inf')
            for duration, cost in zip(durations, costs):
                j = day_index
                while j < len(days) and days[j] < duration + days[day_index]:
                    j += 1
                cache[day_index] = min(cache[day_index], cost + min_cost(j))

            return cache[day_index]

        return min_cost(0)
